etna/lib/etna/auth.rb

```diff
@@ -172,5 +172,15 @@ def approve_hmac(request)
 
       return true
     end
+
+    def encode_path(path)
+      # request.path is pulled from the webserver. In our case, Apache via
+      #   ProxyPass.
+      # However, Apache mod_proxy encodes this differently than URI.encode_www_form_component ... so we normalize it here according to how Etna::Route.path() behaves, so that Metis HMAC signing for
+      # upload / download URLs is compatible.
+      path.split("/").map do |c|
+        URI.encode_www_form_component(URI.decode_www_form_component(c))
```

Comment:
Hmm, I would really worry about how this behaves in cases we aren't testing explicitly here... For example if there are encoded-looking things in the original path this would not work. E.g., if there are two files in a dir, "hey/what+up" and "hey/what up", both of which should be able to exist as independent files in Metis, these both get encoded to the same path by `encode_path`.

Reply:
Ah, so that's an interesting point! Hm, I will have to test and play around with this. I actually expect this branch would correctly protect all the special characters (like `+`).

```
> URI.encode_www_form_component("file+example.txt")
=> "file%2Bexample.txt"
```

So the `+` should get encoded and protected, and when you call `URI.decode_www_form_component()`, it converts back to a `+` instead of a space? I guess it does depend on if `mod_proxy` passes the `+` through as `%2B` or `+`, in which case we might have to resort to using the `nocanon` flag on Apache, because we have no way to tell in the application if it's supposed to be a `+` or a space.

Note that in production, you currently cannot upload a file like `hey/what+up` because of the same problem as the `(` and `)` -- the `+` gets percent-encoded multiple times and results in a 403 from Metis upon calling `upload#start` due to an HMAC mis-match. A file with spaces in the file name works fine. So while we may want such files to independently exist, they currently cannot be uploaded as such -- can only exist if you edit the files post-uploading.